RocketChat/Rocket.Chat
# feat: Allow managing association to business units on departments' creation and update

## Proposed changes (including videos or screenshots)

 - Added a new optional `departmentUnit` object param to `livechat:saveDepartment` Meteor method, `livechat/department` (**POST**) endpoint and `livechat/department` (**PUT**) endpoint;
     - Providing an empty object (`{}`) to this param removes the department from its associated unit (or just doesn't add the department to any unit, if provided on the creation of a new department). Providing `{ _id: undefined }` or `null` triggers the same effect;
     - If no value is provided to this new param, then the association between the department and its unit (if it exists) won't be changed, nor created or removed;
     - When providing a string `_id` field in the object (`{ _id: "unit-id" }`), the department will be associated to the unit with the specified `_id` **IF** the user has the permission to do so.
 - In order to use any of the updated endpoints, the user must have the `manage-livechat-departments` permission (this is already the current behavior), but there are additional conditions to associate a department with a specific unit when the new param is provided:
     - If the user is a **monitor** (role `livechat-monitor`), then they must be a supervisor of the specified unit. That is, they must be linked (as **monitor**) to the specified unit in order to add or remove departments from it;
     - If the user is not a monitor (such as an `admin` or `livechat-manager`), any unit can be associated to the department (on creation or update).
 - Providing any value different from a string to the `_id` field to the new param's object will trigger an error.
 - This is an **additional**/non-blocking operation. No errors are thrown (yet) even if the user tries to add a department to a unit they don't have the permission to manage.

## Issue(s)


## Steps to test or reproduce


Example request to the `livechat:saveDepartment` method (the new param is the last one that should be provided -- `{\"_id\":\"unit-id\"}` in the example):
```
curl --request POST \
  --url http://localhost:3000/api/v1/method.call/livechat%3AsaveDepartment \
  --header 'Accept: application/json' \
  --header 'Content-Type: application/json' \
  --header 'X-Auth-Token: f4U57vG-drmhd8hZbKAT_9TTJqW48Y_VB1zle6HWiRf' \
  --header 'X-User-Id: WM8KyWfu6qXDCRZrq' \
  --data '{"message":"{\"msg\":\"method\",\"id\":\"15\",\"method\":\"livechat:saveDepartment\",\"params\":[null,{\"enabled\":false,\"name\":\"My new department5\",\"description\":\"\",\"showOnRegistration\":false,\"showOnOfflineForm\":false,\"requestTagBeforeClosingChat\":false,\"email\":\"my-new-testdep5@example.com\",\"chatClosingTags\":[],\"offlineMessageChannelName\":\"\",\"abandonedRoomsCloseCustomMessage\":\"\",\"waitingQueueMessage\":\"\",\"departmentsAllowedToForward\":[],\"fallbackForwardDepartment\":\"\",\"allowReceiveForwardOffline\":false},[],{\"_id\":\"unit-id\"}]}"}'
```
The command above triggers the creation of a new department associated with the unit with id `unit-id`. Change the first `null` param to a valid `department` string to edit it.

Example request to the `livechat/department` POST endpoint (check the new `departmentUnit` param):

```
curl --request POST \
  --url http://localhost:3000/api/v1/livechat/department \
  --header 'X-Auth-Token: f4U57vG-drmhd8hZbKAT_9TTJqW48Y_VB1zle6HWiRf' \
  --header 'X-User-Id: WM8KyWfu6qXDCRZrq' \
  --data '{
  "department": {
    "name": "Test23456",
    "enabled": true,
    "showOnOfflineForm": true,
    "showOnRegistration": true,
    "email": "bla@bla"
  },
  "departmentUnit": { "_id": "unit-id" }
}'
```

Example request to the `livechat/department/:id` PUT endpoint (check the new `departmentUnit` param -- providing an empty object to it will remove the department from its associated unit):

```
curl --request PUT \
  --url http://localhost:3000/api/v1/livechat/department/666cc15cd32e4e47b82332fd \
  --header 'X-Auth-Token: f4U57vG-drmhd8hZbKAT_9TTJqW48Y_VB1zle6HWiRf' \
  --header 'X-User-Id: WM8KyWfu6qXDCRZrq' \
  --data '{
  "department": {
    "name": "Test23456",
    "enabled": true,
    "showOnOfflineForm": true,
    "showOnRegistration": true,
    "email": "bla@bla"
  },
  "departmentUnit": {}
}'
```

## Further comments

[SUP-613](https://rocketchat.atlassian.net/browse/SUP-613)

[SUP-613]: https://rocketchat.atlassian.net/browse/SUP-613?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
